MathlibTest/norm_numI.lean

```diff
@@ -0,0 +1,94 @@
+import Mathlib.Analysis.Normed.Ring.Basic
+import Mathlib.Data.Complex.Basic
+import Mathlib.Tactic.NormNum.NormNumI
+import Mathlib.Data.ZMod.Basic
+
+open Complex ComplexConjugate Real Qq
+
+example : (1:ℂ) = 1 + 0 * I := by norm_num1
+example : (I:ℂ) = 0 + 1 * I := by norm_num1
+example : (1.5:ℂ) = 3 / 2 + 0 * I := by norm_num1
+example : (1.0:ℂ) + 0 = 1 := by norm_num1
+example : (1.0:ℂ) + 0.5 = 3/2 := by norm_num1
+example : I + (3/2:ℂ) = 3/2 + I := by norm_num1
+example : 1 * (2:ℂ) = 2 := by norm_num1
+example : (1 + I) * (1 + I * I * I) = 2 := by norm_num1
+example : (1 + 3.5 + I) * (1 + I) = 7 / 2 + 11 / 2 * I := by norm_num1
+example : (3 + 4.5 * I)⁻¹ * (3 + 4.5 * I) = 1 := by norm_num1
+example : -1 / (1 + I) = (I - 1) / 2 := by norm_num1
+example : (I:ℂ) = 0 + 1 * I := by norm_num1
+example : (1.5:ℂ) = ⟨3 / 2, 0⟩ := by conv_lhs => norm_numI
+example : 0 + (1:ℂ) = 1 := by norm_num1
+example : (1.0:ℂ) + 0 = 1 := by norm_num1
+example : (1.0:ℂ) + 0.5 = 3/2 := by norm_num1
+example : I + (3/2:ℂ) = 3/2 + I := by norm_num1
+
+example : I + (3/2:ℂ) = 3/2 + I := by
+  conv_lhs => norm_numI
+  conv_rhs => norm_numI
+
+example : 2 * (2.5:ℂ) = 5 := by norm_num1
+
+example : (1 + I) * (1 + I * I * I) = 2 := by norm_num1
+
+example : (1 + I) * (1 + I * I * I) = 2 := by norm_num1
+
+example : (1 + I) * (1 + I * I * I) = 2 := by
+  conv_lhs => norm_numI
+  conv_rhs => norm_numI
+
+example : (1 + 3.5 + I) * (1 + I) = 7 / 2 + 11 / 2 * I := by norm_num1
+example : (3 + 4 * I)⁻¹ * (3 + 4 * I) = 1 := by norm_num1
+example : -1 / (1 + I) = (I - 1) / 2 := by norm_num1
+example : (1 + I) * (1 - I) = 2 := by norm_num1
+example : (1 + 2 * I) - (1 + 2 * I) = 0 := by norm_num1
+example : (conj (3 + 4 * I) : ℂ) * (3 + 4 * I) = 25 := by norm_num1
+example : (3 + I : ℂ) ^ 2 = 8 + 6 * I := by norm_num1
+example : (3 : ℂ) ^ 2 + (4 : ℂ) ^ 2 = 25 := by norm_num1
+
+example : 3 + I ≠ I ^ 2 := by norm_num1
+example : I ^ 2 ≠ 3 := by norm_num1
+example : 1 + I ≠ 0 := by norm_num1
+example : 1 + I ≠ 1 + 2 * I := by norm_num1
+
+-- This is already pretty slow, but is 2x slower without the `binaryRec` implementation.`
+example : I ^ 11 = -I := by norm_num
```

Comment:
I think this flags that we're using the wrong approach; we should normalize at each step, rather than producing a huge expression and then normalizing it. That would look like storing a `NormNum.Result` for the real and imaginary part, and using `evalMul` etc to combine them under operations.

Reply:
I think `Mathlib.Meta.NormNum.evalMul.core` is supposed to fill that task, through the signature isn't quite right.

Reply:
That does the multiplication, but then you need the second multiplication and the subtraction, and the same again for the imaginary part! So you end up with six lines (and the six lines are very cumbersome because of all the info you need to pass to `Mathlib.Meta.NormNum.evalMul.core` and its friends).

Reply:
Having said that, if we want to refactor the API to allow that, probably we want to land #9915 first.

Reply:
Do you have some tricks in mind to make it look more like `(<- hx1.mul hx2).sub (<- hy1.mul hy2)`?  I do acknowledge that our initial implementation is much faster than the current implementation, but I don't think it's worth *so* much boilerplate.

Reply:
I suspect such a refactor is within reach, but I'd rather not attempt it until #9915 lands, which such a refactor would clash with.